ntoskrnl/ps/quota.c

```diff
@@ -143,45 +519,257 @@ PspInsertQuotaBlock(
     KeReleaseSpinLock(&PspQuotaLock, OldIrql);
 }
 
+/**
+ * @brief
+ * De-references a quota block when quotas
+ * have been returned back because of an
+ * object de-allocation or when a process
+ * gets destroyed. If the last instance
+ * that held up the block gets de-referenced
+ * the function will perform a cleanup against
+ * that block and it'll free the quota block
+ * from memory.
+ *
+ * @param[in] Process
+ * A pointer to a process that de-references the
+ * quota block.
+ *
+ * @param[in] QuotaBlock
+ * A pointer to a quota block that is to be
+ * de-referenced. This block can come from a
+ * process that references it or an object.
+ *
+ * @return
+ * Nothing.
+ */
 VOID
 NTAPI
-PspDestroyQuotaBlock(PEPROCESS Process)
+PspDereferenceQuotaBlock(
+    _In_opt_ PEPROCESS Process,
+    _In_ PEPROCESS_QUOTA_BLOCK QuotaBlock)
 {
-    PEPROCESS_QUOTA_BLOCK QuotaBlock = Process->QuotaBlock;
+    ULONG PsQuotaTypeIndex;
     KIRQL OldIrql;
 
+    /* Make sure the quota block is not trash */
+    ASSERT(QuotaBlock);
+
+    /* Iterate over the process quota types if we have a process */
+    if (Process)
+    {
+        for (PsQuotaTypeIndex = PsNonPagedPool; PsQuotaTypeIndex < PsQuotaTypes; PsQuotaTypeIndex++)
+        {
+            /*
+             * We need to make sure that the quota usage
+             * uniquely associated with the process is 0
+             * on that moment the process gets destroyed.
+             */
+            ASSERT(Process->QuotaUsage[PsQuotaTypeIndex] == 0);
+        }
+
+        /* As the process is now gone, decrement the process count */
+        InterlockedDecrementSizeT(&QuotaBlock->ProcessCount);
+    }
+
+    /* If no one is using this block, begin to destroy it */
     if (QuotaBlock != &PspDefaultQuotaBlock &&
-        InterlockedDecrementUL(&QuotaBlock->ReferenceCount) == 0)
+        InterlockedDecrementSizeT(&QuotaBlock->ReferenceCount) == 0)
```

Comment:
`ReferenceCount` seems to be an `ULONG`, so I think this was correct?
Same for `ProcessCount` above in L571